Migration of the module account_credit_control for Odoo v10

sylvainvh opened the pull request.

Commit:
[MIG] Migration of the module account_credit_control (from v8 to v10)

(cherry picked from commit 811310fd06ca8589fa8e87da6b8bdaea403bf660)

[MIG] Improve the code and fix some bugs

(cherry picked from commit 5bcdfe3f748884d650120f76d1642b820d0a800b)

[FIX] Change the parent for the menu Credit Control (the previous parent menu was not in the Odoo standard)

(cherry picked from commit 826257ce7b5248956d838cae9c94a012d550327e)

[FIX] Fix a lot of errors to respect pep8

(cherry picked from commit bed6cf35d665eea32122c91d33322af141053f29)

[FIX] Respect PEP8

(cherry picked from commit 6353eef0dce1bfe5db8a38d3410d1cc7f833fa10)

pedrobaeza commented:
We are already migrating account_followup module. Why not using it?

sylvainvh commented:
[edit] my bad ! It's a pull request. I'll analyze this module

@pedrobaeza it's a good question ! where I can find this module ? I don't find this module in the repo account-financial-tools

sylvainvh commented:
@pedrobaeza I don't understand the difference between the module account_credit_control and the module account_followup ? It seems to be the same. Right ?

pedrobaeza commented:
That's the question. There isn't difference! The migration is hold until 9.0 migration is completed, but 10.0 migration will be straight-forward. Please check https://github.com/OCA/account-financial-tools/pull/450

sylvainvh commented:
For the moment I'll finish my PR (I want green :p ). 
We will see after that if we merge this module in the module account_followup

Commit:
[ADD] Add setup

(cherry picked from commit 94e10527b9ba59987267713612a4693d1c63e020)

adrienpeiffer commented:
@sylvainvh I juste created a PR on your branch to add the setup file

adrienpeiffer commented on account_credit_control/demo/credit_control_demo.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Are these data still useful ? With experience, it is not a good idea to define account.account as demo data as longer as the default chart of account will not be installed if accounts are created.

adrienpeiffer reviewed (changes requested):
Are the tests performed? I don't see any trace in the logs file.

Commit:
[CHG] account_credit_control: remove demo file.

(cherry picked from commit c32d167ac05ebef2256e3cb5919c13e7b9229fdf)

fclementic2c commented:
@sylvainvh functional test ok
:+1: 
Good Job !

sylvainvh commented:
Now I just need some review

pedrobaeza commented:
@cubells please review and check differences with account_followup

cubells commented on account_credit_control/__manifest__.py:
```diff
@@ -19,33 +19,45 @@
 {'name': 'Account Credit Control',
- 'version': '8.0.0.3.0',
- 'author': "Camptocamp,Odoo Community Association (OCA)",
+ 'version': '10.0.0.3.0',
```
Correct version to 10.0.0.1.0

pedrobaeza replied:
Actually 10.0.1.0.0

cubells replied:
It's true @pedrobaeza 

Thanks.

cubells commented on account_credit_control/__manifest__.py:
```diff
@@ -19,33 +19,45 @@
+     'account',
+     'mail',
+ ],
  'website': 'http://www.camptocamp.com',
```
It must be https

cubells commented on account_credit_control/data/data.xml:
```diff
@@ -0,0 +1,249 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="1">
```
It must be: odoo noupdate=1

cubells commented on account_credit_control/__manifest__.py:
```diff
@@ -19,33 +19,45 @@
+     # Security
+     "security/ir.model.access.csv",
+ ],
+ 'demo': [
```
Remove empty keys in dictionary.

cubells commented on account_credit_control/models/account_account.py:
```diff
@@ -18,7 +18,7 @@
 #
 ##############################################################################
-from openerp import models, fields
+from odoo import models, fields
```
Change order: from odoo import fields, models

cubells commented on account_credit_control/models/account_invoice.py:
```diff
@@ -18,7 +18,8 @@
 #
 ##############################################################################
-from openerp import models, fields, api, _
+from odoo import models, fields, api, _
```
Change order: _, api, fields, models

guewen replied:
This is a non-blocking nitpicking, mind you

pedrobaeza replied:
Indeed, but it's a nice to have if you don't mind doing the rest of the comments. If not, as said, don't block the PR for that.

cubells commented on account_credit_control/models/credit_control_line.py:
```diff
@@ -20,9 +20,10 @@
 import logging
 
-from openerp import models, fields, api, _
+from odoo import models, fields, api, _
```
Same as above

cubells commented on account_credit_control/models/credit_control_policy.py:
```diff
@@ -18,7 +18,8 @@
 #
 ##############################################################################
-from openerp import models, fields, api, _
+from odoo import models, fields, api, _
```
same as above

cubells commented on account_credit_control/models/credit_control_run.py:
```diff
@@ -20,9 +20,10 @@
 import logging
 
-from openerp import models, fields, api, _
+from odoo import models, fields, api, _
```
same as above

cubells commented on account_credit_control/views/credit_control_line.xml:
```diff
@@ -0,0 +1,180 @@
+                    <filter name="filter_draft" icon="terp-mail-message-new"
+                            string="Draft" domain="[('state', '=', 'draft')]"
+                            help="Draft lines have to be triaged."/>
+                    <filter name="filter_to_be_sent" icon="terp-mail-forward"
```
You must use font-awesome icons if necessary. Gtk icons are not already available in odoo v.10.

cubells commented on account_credit_control/views/credit_control_line.xml:
```diff
@@ -0,0 +1,180 @@
+        <field name="model">credit.control.line</field>
+        <field name="arch" type="xml">
+            <search string="Control Credit Lines">
+                <group string="Filters">
```
Add name selector if you can, as it is recommended at https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0

cubells commented on account_credit_control/views/credit_control_line.xml:
```diff
@@ -0,0 +1,180 @@
+                </group>
+
+                <newline/>
+                <group expand="0" string="Group By...">
```
Add name selector if you can, as it is recommended at https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0

cubells commented on account_credit_control/wizard/credit_control_communication.py:
```diff
@@ -19,9 +19,9 @@
 ##############################################################################
 import logging
-from openerp import models, fields, api
+from odoo import models, fields, api
```
same as above

cubells reviewed (changes requested):
Code review

cubells commented:
@sylvainvh 

README file information is incorrect, outdated and incomplete.

Can you change README file in order to complete Usage and Configuration information about module?

And README file must follow this oca style: https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

cubells commented on account_credit_control/wizard/credit_control_policy_changer.py:
```diff
@@ -52,18 +55,17 @@ def _get_default_lines(self):
 
             domain = [('account_id', '=', invoice.account_id.id),
                       ('move_id', '=', invoice.move_id.id),
                       ('reconcile_id', '=', False)]
```
It must be field reconciled instead of reconcile_id

sylvainvh replied:
The field "reconcile" doesn't exist on the model account.move.line. I think it should be "full_reconcile_id". What do you think @cubells ?

cubells replied:
Correct field, as I said above, is "reconciled". 

Field "full_reconcile_id" is a Many2one field for store reconcile_id when reconcile is full. This field is created by module "account_full_reconcile", that is not in dependencies.

sylvainvh replied:
Oké done on the last commit

cubells commented on account_credit_control/security/ir.model.access.csv:
```diff
@@ -2,7 +2,7 @@
 "account_credit_control.ir_model_access_270",1,1,"group_account_credit_control_manager","credit_control_manager_line","account_credit_control.model_credit_control_line",1,1
```
Can you put fields in order? It's more readable. 

Perm fields later.

cubells commented on account_credit_control/migrations/0.3/post-migration.py:
```diff
@@ -1,7 +1,7 @@
 # -*- encoding: utf-8 -*-
 import logging
```
Post-migration script folder must be 10.0.1.x format not 0.3

pedrobaeza replied:
I bet that this migration script is for older versions and this should be removed.

cubells reviewed (changes requested):
Functionally review

cubells commented:
@sylvainvh :
It would be interesting to have some graphic or pivot view in order to analyze over due.

guewen commented on account_credit_control/models/credit_control_line.py:
```diff
@@ -20,9 +20,10 @@
 
-from openerp import models, fields, api, _
+from odoo import models, fields, api, _
+from odoo.exceptions import Warning
```
The correct error is `UserError` (same remark for the other files)

Commit:
[MIG] Clean the code to respect PEP8 (remove useless imports, remove useless spaces, ...) and fix some small bugs

(cherry picked from commit efd9e9f8efe9a5e8da791d179b5a5d57710a31c3)

[FIX] Add space to respect PEP8

(cherry picked from commit 06c93c610646bb4269f303c3a7a55700cc6dfd1f)

[FIX] Replace the field "full_reconcile_id" by "reconciled" in a search domain on account.move.line

(cherry picked from commit cf2cfd43e34dff9eef2571b87de777a183c220c0)

sylvainvh commented:
I made somes modification. Can you review ?

cubells commented on account_credit_control/README.rst:
```diff
@@ -1,11 +1,16 @@
 Configuration
 -------------
 
 Configure the policies and policy levels in ``Accounting  > Configuration >
```
It's "Invoicing" now, not "Accounting"

pedrobaeza replied:
Maybe you should put both terms. If you only install module _account_, you have "Invoicing". When you install _account_accountant_, then the menu is renamed to "Accounting".

cubells commented on account_credit_control/README.rst:
```diff
@@ -16,14 +21,24 @@ You are able to specify a particular policy for one partner or one invoice.
 Usage
```
There isn't an Accounting tab in company form now. There is a "General information" tab,

cubells replied:
Please add some information here in configuration about accounts tab in every policy because if user doesn't enter any account, policy per partner can't be selected.

cubells commented on account_credit_control/README.rst:
```diff
@@ -16,14 +21,24 @@ You are able to specify a particular policy for one partner or one invoice.
 
-Menu entries are located in ``Accounting > Periodical Processing > Credit
-Control``.
+Menu entries are located in ``Accounting > Report > Credit Control``.
```
"Invoicing" please
"Reports" please

sylvainvh replied:
I don't undersand, the name of the main menu is "Accounting". I don't see the menu "Invoicing"

cubells commented on account_credit_control/README.rst:
```diff
@@ -16,14 +21,24 @@ You are able to specify a particular policy for one partner or one invoice.
+
+* Nicolas Bessi (Camptocamp)
+* Guewen Baconnier (Camptocamp)
+* Sylvain Van Hoof (Okia SPRL) <sylvain@okia.be>
```
You must add Try button:
`
.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
   :alt: Try me on Runbot
   :target: https://runbot.odoo-community.org/runbot/92/10.0`

cubells replied:
You must add "Bug Tracker" section and "Maintainer" section as is indicated in maintainers example:

https://raw.githubusercontent.com/OCA/account-financial-tools/10.0/account_check_deposit/README.rst

cubells commented on account_credit_control/__manifest__.py:
```diff
@@ -19,33 +19,43 @@
- 'version': '8.0.0.3.0',
- 'author': "Camptocamp,Odoo Community Association (OCA)",
+ 'version': '10.0.1.0.0',
+ 'author': "Camptocamp,Odoo Community Association (OCA),Okia",
```
You must separate authors or you get a linter error:

Correct is  "Camptocamp, Odoo Community Association (OCA), Okia",

pedrobaeza replied:
No, that's not an actual problem

cubells commented on account_credit_control/README.rst:
```diff
@@ -1,11 +1,16 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
```
Please, change all headers of all files by this new format:

https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L1-L3

jgrandguillaume commented:
Hi,

@cubells Thank you for your reviews, very much appreciated ! That being said, lot's of things here sounds like a bit picky for this PR if I may...

I've tested it, functionally speaking it it works great, Travis is green, so on my side :+1: to merge it.

Regards,

Joël

pedrobaeza commented:
@jgrandguillaume although picky, we have imposed ourselves these high standards to conform 100% of the OCA guidelines. As we are making the effort to dispose the time spent on account_followup as discussed in https://github.com/OCA/account-financial-tools/pull/450 in favor of this module, and still needing by our part to:

* migrate to v9
* create migration scripts

this is not asking too much IMO.

adrienpeiffer commented on account_credit_control/views/res_partner.xml:
```diff
@@ -0,0 +1,23 @@
+        <field name="arch" type="xml">
+            <field name="credit" position="after">
+                <field name="credit_policy_id" widget="selection"
+                       groups="account_credit_control.group_account_credit_control_manager,account_credit_control.group_account_credit_control_use"/>
```
account_credit_control.group_account_credit_control_use**R**

yvaucher replied:
good catch

adrienpeiffer commented on account_credit_control/views/credit_control_line.xml:
```diff
@@ -0,0 +1,180 @@
+    <menuitem
+            name="Credit Control"
+            parent="account.menu_finance_reports"
+            id="base_credit_control_menu" sequence="10"/>
```
Add security groups here

adrienpeiffer commented on account_credit_control/views/account_invoice.xml:
```diff
@@ -0,0 +1,44 @@
+                    <group>
+                        <field name="credit_policy_id" widget="selection"
+                               string="Manual Credit Control Policy"
+                               attrs="{'invisible': [('credit_policy_id', '=', False)]}"
```
I don't really understand this condition. It means that the field is invisible if this one is empty. So, it will always invisible ?

grindtildeath replied:
The field is readonly, so if it isn't defined on the invoice there's no use to display it. I just removed the widget so it only appears as a link.

adrienpeiffer replied:
@grindtildeath Ok, and why this one is readonly and when it is filled ?
cc\ @sylvainvh

adrienpeiffer commented on account_credit_control/views/credit_control_run.xml:
```diff
@@ -0,0 +1,73 @@
+                    <button name="generate_credit_lines"
+                            string="Compute Credit Control Lines"
+                            class="oe_highlight"
+                            type="object" icon="gtk-execute"
```
Change this one to fontawesome icon. e.g. fa-cog

adrienpeiffer commented:
The PDF report isn't very well proportionate

![acc2](https://cloud.githubusercontent.com/assets/8035793/26546262/63c18a64-4469-11e7-8e10-b96ab76d4a04.jpg)

grindtildeath committed:
Fix errors, improve menus, views, readme

(cherry picked from commit 93a9b1ee4fd05c2a8fd97c305e392cab6a24f865)

grindtildeath commented:
@sylvainvh 
I did some improvements according to review and will open a PR on your branch as soon as I will have fixed the PDF report.

sylvainvh commented:
Hey, thank you for your feedback !!! I have to take the time to complete this PR.
@grindtildeath Thank you for you PR. I'll check that ASAP (it means now)

grindtildeath commented:
Opened : https://github.com/QANSEE/account-financial-tools/pull/3

adrienpeiffer commented on account_credit_control/views/account_invoice.xml:
```diff
@@ -0,0 +1,44 @@
+                    <group>
+                        <field name="credit_policy_id"
+                               string="Manual Credit Control Policy"
+                               attrs="{'invisible': [('credit_policy_id', '=', False)]}"
```
@grindtildeath why this one is readonly and when it is filled ?
cc\ @sylvainvh

grindtildeath replied:
@adrienpeiffer While testing, I did fill it using the wizard 'credit.control.policy.changer'.
Now I don't know if its purpose is to change an existing policy on an invoice or to define it the first time. Maybe @fclementic2c or @jgrandguillaume who did functional tests can let us know.

adrienpeiffer replied:
Ok, I didn't catch the use of this wizard. So, it's ok for me :smile:

grindtildeath commented:
@adrienpeiffer While testing, I did fill it using the wizard 'credit.control.policy.changer'. 
Now I don't know if its purpose is to change an existing policy on an invoice or to define it the first time. Maybe @fclementic2c or @jgrandguillaume who did functional tests can let us know.

sylvainvh commented:
@grindtildeath I think there is a problem with the last PR. I'll try to see where is the problem

Commit:
[MIG] Add some unit tests on the module account_credit_control

(cherry picked from commit 7fa1b0ac883914f3cdfcaaa1a00016a35cced368)

[IMP] Add some unitest and fix a bug with exceptions

(cherry picked from commit c2f34ecd06fa5ba093084cf32f35c722c5866ff5)

[FIX] Reindent line

(cherry picked from commit b643a48f86b20ed6496f2e08366a407095d1540b)

[IMP] Add some new unittests

(cherry picked from commit f118e2391de3b443558d26186fe44f55a7f0b9c4)

[IMP] Add encoding in each python files

(cherry picked from commit 9599a88303b921c8e73174db61318f6dc9f02017)

sylvainvh commented:
Fixed :p (or not)

grindtildeath commented:
@adrienpeiffer It seems there was an issue with my user record on OCA website. Should be fixed now, even if I don't know why it suddenly appears here after having already contributed to many PRs :smile:

sylvainvh commented:
Youpiiii !!! Everything work. 
We can merge it now ? :p @pedrobaeza

benwillig commented:
Functionnal test ok & code review :+1:

cubells reviewed (approved):
:+1: 
LGTM

good job @sylvainvh

pedrobaeza commented:
One last thing to do before merging: please squash commits by logical set/author.

sylvainvh commented:
Here we are ! I hope it should be oke for you @pedrobaeza  :) 
I squashed all my commits by "features" (migration, improvement, unittest).

pedrobaeza commented:
Congratulation for the hard work and let's welcome the credit control module in v10!
